File: javascript/rle.js
// run length encoding; simple data compression technique that replaces repeated sequences 
// of identical characters with a single instance of the character followed by a count

function encode(str)
{
  let encodedString = "";
  let count = 1;

  for(let i = 0; i < str.length; i++)
    {
      if(i+1 < str.length && str[i] === str[i+1])
      {
        count++;
      }
      else
      {
        encodedString += str[i] + count;
        count = 1;
      }
    }
  return encodedString;
}

function decode(str)
{
  let decodedString = "";
  let i = 0;

  while(i < str.length)
    {
      let count = "";
      while(i < str.length && !isNan(parseInt(str[i])))
        {
          count += str[i];
          i++;
        }
      decodedString += str[i].repeat(parseInt(count));
      i++;
    }
  return decodedString;
}
// example usage
const inputString = "AAAABBBCCDAA";
const encodedString = encode(inputString);
const decodedString = decode(encodedString);

console.log("original string:", inputString);
console.log("encoded string:", encodedString);
console.log("decoded string:", decodedString); 

File: cpp/RLE.cpp
// replaces consecutive occurences of same character with a single instance of the character and the run length
// effective when data contains many runs of identical characters

#include <iostream>
#include <string>
#include <vector>
#include <sstream>

std::string rle_encode(const std::string& text) {
  std::stringstream encdoed;
  if(text.empty()) return "";

  char current_char = text[0];
  int current_count = 1;

  for(size_t i = 1; i < text.length(); ++i) {
    if(text[i] == current_char) {
      current_count++;
    } else {
      encoded << current_char << current_coutn;
      current_char = text[i];
      current_count = 1;
    }
  }
  encoded << current_char << current_count;
  return encoded.str();
}

std::string rle_decode(const std::string& encoded) {
  std::stringstream decoded;
  if(encoded.empty()) return "";

  for(size_t i = 0; i < encoded.length(); i += 2) {
    char ch = encoded[i];
    int count = std::stoi(encoded.substr(i+1, 1));

    for(int j = 0; j < count; ++j) {
      decoded << ch;
    }
  }
  return decoded.str();
}

int main() {
  std::string text = "AAABBCDDDDEEEE";
  std::string encoded = rle_encode(text);
  std::string decoded = rle_decode(encoded);

  std::cout << "original: " << text << std::endl;
  std::cout << "encoded: " << encoded << std::endl;
  std::cout << "decoded: " << decoded << std::endl;

  return 0;
}
